src/spatial/utils.jl

```diff
@@ -0,0 +1,219 @@
+# or just don't update the target site if it's not a valid site
+
+################### abstract spatial rates struct ###############
+
+abstract type AbstractSpatialRates end
+
+#TODO refactor names of functions?
+
+#return rate of site
+function get_site_rate(spatial_rates_struct, site_id)
+    get_site_reactions_rate(spatial_rates_struct,site_id)+get_site_diffusions_rate(spatial_rates_struct,site_id)
+end
+
+#return rate of reactions at site
+function get_site_reactions_rate end
+
+#return rate of diffusions at site
+function get_site_diffusions_rate end
+
+#returns an iterator over the pairs (rx_id,rx_rate) of a site
+function get_site_reactions_iterator end
+
+#returns an iterator over the pairs (species_id, diffusion_rate) of a site
+function get_site_diffusions_iterator end
+
+#sets the rate of reaction at site
+function set_site_reaction_rate! end
+
+#sets the rate of diffusion at site
+function set_site_diffusion_rate! end
+
+#################### SpatialRates <: AbstractSpatialRates ######################
+
+# NOTE: for now assume diffusion rate only depends on the site and species (and not the neighbor of the site)
+# TODO make these matrices instead of vectors of vectors for better performance
+# TODO refactor names
+# TODO add doc strings to explain what these are
+struct SpatialRates{R} <: AbstractSpatialRates
+    reaction_rates::Vector{Vector{R}} # [[reaction rates for site 1],...,[reaction rates for last site]]
+    diffusion_rates::Vector{Vector{R}} # [[diffusion rates for site 1],...,[diffusion rates for last site]]
+    reaction_rates_sum::Vector{R} # [sum of reaction rates for site 1,...,sum of reaction rates for last site]
+    diffusion_rates_sum::Vector{R} # [sum of diffusion rates for site 1,...,sum of diffusion rates for last site]
+end
+
+"""
+standard constructor, assumes numeric values for rates
+"""
+function SpatialRates(reaction_rates::Vector{Vector{R}}, diffusion_rates::Vector{Vector{R}}) where {R}
+    SpatialRates{R}(reaction_rates, diffusion_rates, [sum(site) for site in reaction_rates], [sum(site) for site in diffusion_rates])
+end
+#QUESTION are the type annotations here correct? Do I need them at all?
+"""
+initializes SpatialRates with zero rates
+"""
+function SpatialRates(num_jumps::Integer,num_species::Integer,num_sites::Integer)
+    reaction_rates = [zeros(Float64, num_jumps) for i in 1:num_sites]
+    diffusion_rates = [zeros(Float64, num_species) for i in 1:num_sites]
+    SpatialRates(reaction_rates,diffusion_rates)
+end
+
+"""
+initializes SpatialRates with zero rates
+"""
+function SpatialRates(ma_jumps, num_species, spatial_system::AbstractSpatialSystem)
+    num_sites = number_of_sites(spatial_system)
+    num_jumps = get_num_majumps(ma_jumps)
+    SpatialRates(num_jumps,num_species,num_sites)
+end
+
+"""
+return rate of reactions at site
+"""
+function get_site_reactions_rate(spatial_rates, site_id)
```

Comment:
maybe call `total_site_rx_rate` or `total_site_rxrate`? Similar for the diffusion rate.